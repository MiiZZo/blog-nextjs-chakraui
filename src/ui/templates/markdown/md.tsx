import React, { useRef, useEffect } from "react";
import MD, { MarkdownOptions } from "markdown-to-jsx";
import hljs from "highlight.js";
import { Heading, Text, List, ListItem, PseudoBox } from "@chakra-ui/core";
import { Blockquote } from "@ui";

interface MarkdownProps {
  children: string;
}

export function Markdown({ children }: MarkdownProps) {
  const rootRef = useRef<HTMLDivElement>();

  useEffect(() => {
    rootRef.current
      .querySelectorAll("pre code")
      .forEach((block: HTMLElement) => {
        hljs.highlightBlock(block);
      });
  }, [children]);

  return (
    <div ref={rootRef}>
      <MD options={mdOptions}>{children}</MD>
    </div>
  );
}

const mdOptions: MarkdownOptions = {
  overrides: {
    h1: {
      component: Heading,
      props: { fontSize: "2xl", mb: 5 }
    },
    h2: {
      component: Heading,
      props: { fontSize: "xl", mb: 5 }
    },
    h3: {
      component: Heading,
      props: { fontSize: "lg", mb: 5 }
    },
    h4: {
      component: Heading,
      props: { fontSize: "md", mb: 5 }
    },
    h5: {
      component: Heading,
      props: { fontSize: "sm", mb: 5 }
    },
    h6: {
      component: Heading,
      props: { fontSize: "xs", mb: 5 }
    },
    p: {
      component: Text,
      props: { mb: 5 }
    },
    li: {
      component: ListItem,
      props: { mb: 5 }
    },
    ul: {
      component: List,
      props: { styleType: "disc" }
    },
    ol: {
      component: List,
      props: { styleType: "decimal" }
    },
    code: {
      component: PseudoBox,
      props: { as: "code", mb: 5 }
    },
    blockquote: {
      component: Blockquote
    }
  }
};
