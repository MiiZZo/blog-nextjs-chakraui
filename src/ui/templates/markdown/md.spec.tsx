import React from "react";
import { cleanup } from "@testing-library/react";
import { myRender } from "@lib/testing/render";
import { Markdown } from "./md";

afterEach(cleanup);

describe("md component", () => {
  it("should match a shapshot", () => {
    const { asFragment } = myRender(
      <Markdown>
        {`
          # My blockquote
  
          \`\`\`jsx
            import React from "react";
  
            const Component = () => <div></div>
          \`\`\`
        `}
      </Markdown>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
