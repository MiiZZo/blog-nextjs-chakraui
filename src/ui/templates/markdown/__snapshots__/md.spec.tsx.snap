// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`md component should match a shapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <pre>
        <code
          class="css-1e0ioh3 hljs 1c"
        >
                
          <span
            class="hljs-meta"
          >
            # My blockquote
          </span>
          
  
      
        </code>
      </pre>
      <pre>
        <code
          class="lang-jsx css-1e0ioh3 hljs javascript"
        >
                      
          <span
            class="hljs-keyword"
          >
            import
          </span>
           React 
          <span
            class="hljs-keyword"
          >
            from
          </span>
           
          <span
            class="hljs-string"
          >
            "react"
          </span>
          ;
  
            
          <span
            class="hljs-keyword"
          >
            const
          </span>
           Component = 
          <span
            class="hljs-function"
          >
            () =&gt;
          </span>
           
          <span
            class="xml"
          >
            <span
              class="hljs-tag"
            >
              &lt;
              <span
                class="hljs-name"
              >
                div
              </span>
              &gt;
            </span>
            <span
              class="hljs-tag"
            >
              &lt;/
              <span
                class="hljs-name"
              >
                div
              </span>
              &gt;
            </span>
          </span>
        </code>
      </pre>
    </div>
  </div>
</DocumentFragment>
`;
